### PR TITLE
[corlib] Type from CoreFX

### PR DESCRIPTION
**Pull request by MaximLipnin**
Part of #7737

### CODE REVIEW

**Comment by MaximLipnin** [comment]
https://github.com/mono/mono/pull/11725 needs to be merge to fix `GetTypeByName_Invalid` test which was re-enabled a bit early in https://github.com/mono/corefx/pull/187.

**Comment by MaximLipnin** [comment]
`CustomAttributes` test from https://github.com/mono/mono/issues/11569 was re-enabled in https://github.com/mono/corefx/pull/186 and doesn't fail anymore.

**Comment by lambdageek** [comment]
@MaximLipnin Do you know if there are some fairly extensive tests of https://github.com/mono/corefx/blob/3931d3640bd3e08bc6cd8bddee27a832ddebd896/src/Common/src/CoreLib/System/Type.cs#L357-L362

It's bringing in the whole `SignatureType` user defined `Type` hierarchy, and without pretty good tests I'm worried we won't have a good idea of how the Mono runtime will deal with them.

**Comment by MaximLipnin** [comment]
@lambdageek [SignatureTypes.netcoreapp.cs](https://github.com/mono/corefx/blob/master/src/System.Runtime/tests/System/Reflection/SignatureTypes.netcoreapp.cs) fixture contains set of tests for `MakeGenericMethodParameter`. Should I enable them too?

**Comment by MaximLipnin** [comment]
@marek-safar There is extra work related to `Type.MakeGenericMethodParameter` method and `SignatureType` classes. Should we proceed with it in this PR?

**Comment by marek-safar** [comment]
@MaximLipnin yes, it'd be nice to land them together

File: mcs/class/corlib/corert/ReflectionAugments.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+using System;
+using System.Reflection;
+
+namespace Internal.Reflection.Augments {
```
Comment (lambdageek) [inline mcs/class/corlib/corert/ReflectionAugments.cs]: We don't normally use the `Internal` namespace in Mono.  Also i'm not sure what this has to do with `corert`

File: mcs/class/referencesource/mscorlib/system/rttype.cs [hunk 2]
```diff
@@ -4258,6 +4258,10 @@ public override Type MakeGenericType(Type[] instantiation)
 
                 if (rtInstantiationElem == null)
                 {
+#if MONO                    
+                    if (instantiationElem.IsSignatureType)
+                        return Internal.Reflection.Augments.ReflectionAugments.MakeGenericSignatureType (this, instantiation);
```
Comment (lambdageek) [inline mcs/class/referencesource/mscorlib/system/rttype.cs]: Shouldn't mono also return a `TypeBuilderInstantiation`?

File: mcs/class/corlib/corert/ReflectionAugments.cs (hunk 1 above)
Comment (MaximLipnin) [reply mcs/class/corlib/corert/ReflectionAugments.cs]: If we don't want to bring the whole classes from CoreRT we just put some pieces to `corert` preserving original namespace. E.g. the same is done in `corert/RuntimeAugments.cs`.

File: mcs/class/referencesource/mscorlib/system/rttype.cs (hunk 2 above)
Comment (MaximLipnin) [reply mcs/class/referencesource/mscorlib/system/rttype.cs]: I've just found that there is the difference between dotnet/corefx and mono/corefx.
`Type` in upstream contains [MakeGenericSignatureType ](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Type.cs#L349)method and Mono fork doesn't. So I need to update `Type` in mono/corefx.

**Comment by MaximLipnin** [comment]
@monojenkins build failed

**Comment by MaximLipnin** [comment]
Most of lanes are green. Please review.

**Comment by marek-safar** [comment]
@monojenkins commit apidiff and csproj

**Comment by MaximLipnin** [comment]
@monojenkins build failed

**Comment by MaximLipnin** [comment]
@monojenkins build Windows x64 C++

**Comment by MaximLipnin** [comment]
@monojenkins build Windows i386

**Comment by marek-safar** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins build Linux x64 All Profiles Build

File: mcs/class/referencesource/mscorlib/system/rttype.cs [hunk 3]
```diff
@@ -4259,6 +4264,10 @@ public override Type MakeGenericType(Type[] instantiation)
 
                 if (rtInstantiationElem == null)
                 {
+#if MONO                    
+                    if (instantiationElem.IsSignatureType)
+                        return MakeGenericSignatureType (this, instantiation);
```
[the remark this answers is not in the document]
Comment (MaximLipnin) [reply mcs/class/referencesource/mscorlib/system/rttype.cs]: Yes, but constructor of `SignatureConstructedGenericType` produced by `Type.MakeGenericSignatureType` method has this null-check. https://github.com/dotnet/corefx/blob/664d98b3dc83a56e1e6454591c585cc6a8e19b78/src/Common/src/CoreLib/System/Reflection/SignatureConstructedGenericType.cs#L24-L28

**Pull request merged**